primitives/src/vstaging/executor_params.rs

```diff
@@ -0,0 +1,124 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Abstract execution environment parameter set.
+//!
+//! Parameters are encoded as (u8, u64) tuples, where u8 is a tag defining a type of
+//! parameter, and u64 is a parameter value. Parameter set and value structure are opaque
+//! and depend on execution environment itself. Decoding to a usable semantics structure is
+//! done in `polkadot-node-core-pvf`.
+
+use parity_scale_codec::{Decode, Encode};
+use parity_util_mem::MallocSizeOf;
+use scale_info::TypeInfo;
+use sp_std::{ops::Deref, vec::Vec};
+
+/// Execution environment type
+#[derive(Clone, Copy, Debug, Encode, Decode, PartialEq, Eq, MallocSizeOf, TypeInfo)]
+pub enum ExecutionEnvironment {
+	/// Generic Wasmtime executor
+	WasmtimeGeneric = 0,
+}
+
+/// Executor instantiation strategy
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, MallocSizeOf, TypeInfo)]
+pub enum ExecInstantiationStrategy {
+	/// Pooling copy-on-write
+	PoolingCoW,
+	/// Recreate instance copy-on-write
+	RecreateCoW,
+	/// Pooling
+	Pooling,
+	/// Recreate instance
+	Recreate,
+	/// Legacy instantiation strategy
+	Legacy,
+}
+
+/// A single executor parameter
+#[non_exhaustive]
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, MallocSizeOf, TypeInfo)]
+pub enum ExecutorParam {
+	/// General parameters:
+	/// Execution environment type
+	Environment(ExecutionEnvironment),
+	/// Version of the set of execution parameters, unique in the scope of execution environment type
+	Version(u32),
+
+	/// Parameters setting the executuion environment semantics:
+	/// Number of extra heap pages
+	ExtraHeapPages(u64),
+	/// Max. memory size
+	MaxMemorySize(u32),
+	/// Wasm logical stack size limit, in units
+	StackLogicalMax(u32),
+	/// Executor machine stack size limit, in bytes
+	StackNativeMax(u32),
+	/// Executor instantiation strategy
+	InstantiationStrategy(ExecInstantiationStrategy),
+	/// `true` if compiler should perform NaNs canonicalization
+	CanonicalizeNaNs(bool),
+	/// `true` if parallel compilation is allowed, single thread is used otherwise
+	ParallelCompilation(bool),
+}
+
+/// Deterministically serialized execution environment semantics
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, MallocSizeOf, TypeInfo)]
+pub struct ExecutorParams(Vec<ExecutorParam>);
```

Comment:
nit: Why is this a vec? 1. Why do we want a deterministically-serialized data structure, and 2. can we not use a struct with the [`repr(c) annotation`](https://github.com/rust-lang/reference/blob/8e7d614303b0dec7492e048e63855fcd3b944ec8/src/type-layout.md#representations)?

Reply:
1) As a part of on-chain storage, it is participating in consensus, so it has to be deterministic
2) We don't want to just replicate `Semantics` as we want to be able to use it for different execution environments (which can use completely different semantics) and for some future versions of Wasmtime as well (which can add some new fields to `Semantics`). If it were a structure, we would need to do a storage migration every time when something new is added. Being just a vector, it's extensible without bound.

Besides that, we don't want to keep a full EE semantics for every session. We have a [default EE configuration](https://github.com/paritytech/polkadot/pull/6161/files#diff-f4ffe6a7003b388a382a12c2ac95731f8862f4e3cd96be2898f271c131ee528aR52) and `ExecutorParams` only keeps changes to it. So an empty vector is a perfectly legit `ExecutorParams` which only means to use all the default ones. It helps to keep volume of on-chain data low.

Reply:
Thank you for the clear explanation! I think it's worth adding this to the doc for this struct.